src/prefect/runner/runner.py

```diff
@@ -850,18 +889,84 @@ async def _cancel_run(self, flow_run: "FlowRun", state_msg: Optional[str] = None
                     "message": state_msg or "Flow run was cancelled successfully."
                 },
             )
+
+            flow, deployment = await self._get_flow_and_deployment(flow_run)
+            self._emit_flow_run_cancelled_event(
+                flow_run=flow_run, flow=flow, deployment=deployment
+            )
+            run_logger.info(f"Cancelled flow run '{flow_run.name}'!")
+
+    async def _get_flow_and_deployment(
+        self, flow_run: "FlowRun"
+    ) -> tuple[Optional["APIFlow"], Optional["Deployment"]]:
+        deployment: Optional["Deployment"] = (
+            self._deployment_cache.get(flow_run.deployment_id)
+            if flow_run.deployment_id
+            else None
+        )
+        flow: Optional["APIFlow"] = self._flow_cache.get(flow_run.flow_id)
+        if not deployment:
             try:
                 deployment = await self._client.read_deployment(flow_run.deployment_id)
+                self._deployment_cache[flow_run.deployment_id] = deployment
```

Comment:
will this gracefully handle the case where `flow_run.deployment_id` is `None`?

Reply:
Good catch, it probably wouldn't, but should as of [1172502](https://github.com/PrefectHQ/prefect/pull/16410/commits/1172502bed51f58268c19d27f1943aa7033534c4).